src/doc/contrib/src/implementation/schemas.md

```diff
@@ -0,0 +1,36 @@
+# Data Schemas
+
+Cargo reads and writes user and machine facing data formats, like
+- `Cargo.toml`, read and written on `cargo package`
+- `Cargo.lock`, read and written
+- `.cargo/config.toml`, read-only
+- `cargo metadata` output
+- `cargo build --message-format` output
+
+## Schema Design
+
+Generally,
+- Fields should be kebab case
+  - `#[serde(rename_all = "kebab-case")]` should be applied defensively
+- Fields should only be present when needed, saving space and parse time
+  - Also, we can always switch to always outputting the fields but its harder to stop outputting them
+  - `#[serde(skip_serializing_if = "Default::default")]` should be applied liberally
+- For output, prefer [jsonlines](https://jsonlines.org/) as it allows streaming output and flexibility to mix content (e.g. adding diagnostics to output that didn't prevously have it
+- `deny_unknown_fields` should not be used to allow evolution of formats, including feature gating
+
+## Schema Evolution Strategies
```

Comment:
When reading this, each bullet point is a bit disconnected to me.
I can't tell what it means exactly by evolution strategies. Maybe it is a pretty common term but I didn't know?

Also like "When reading data, add new fields", I had a hard time interpreting it as part of evolution strategies. For "Versioned with the `edition`" I am able to guess it means "When a schema needs to change some existing field, versioning it with `edition`", though the whole paragraph still looks pretty sparse and needs some guesses for me to read.

Reply:
Tried to expand on it a bit more.  Hope it helps!

Reply:
Thanks. That is good enough for me as a maintainer. Not sure about casual contributors, but we can always iterate later :)